### ISSUE TITLE
Serialize PL_sv_yes and PL_sv_no as boolean true / false

### ISSUE THREAD

**Issue by dolmen** [rurban/Cpanel-JSON-XS#39]
perl has special values that could be interpreted as booleans: PL_sv_yes and PL_sv_no. They can be created from pure perl using `!0` and `!1`. YAML::XS uses them for boolean round-tripping.
```
$ perl -MDevel::Peek -MO=Concise -e 'Dump([ !0, !1 ])'
8  <@> leave[1 ref] vKP/REFC ->(end)
1     <0> enter ->2
2     <;> nextstate(main 1 -e:1) v:{ ->3
7     <2> Dump vK/1 ->8
6        <@> anonlist sK*/1 ->7
3           <0> pushmark s ->4
4           <$> const(SPECIAL sv_yes) s/FOLD ->5
5           <$> const(SPECIAL sv_no) s/FOLD ->6
-e syntax OK
```

Cpanel::JSON::XS (like JSON::XS) doesn't recognize them:
```
$ $ perl -MCpanel::JSON::XS -E 'print Cpanel::JSON::XS::encode_json([ !0, !1 ])'
[1,""]
```

I expect this output:
```
[true,false]
```

This issue is painful to convert YAML to JSON while keeping booleans values properly typed. See [this StackOverflow question](https://stackoverflow.com/questions/32763676/how-to-preserve-boolean-type-from-yaml-document-in-perl-5).

**Comment by dolmen** [comment]
The more I learn about PL_nv_yes / PL_nv_no, the more I think the issue is in YAML::XS...

**Comment by rurban** [comment]
I' d say that mapping those two makes a lot of sense. 
It might create backcompat problems: 1 might be true, "" might be false then, 
but it's a good 1:1 mapping to core.

**Comment by rurban** [comment]
Merged into master for the next release 3.0201

**Comment by perlpunk** [comment]
It seems !0 and !1 are only special when used directly. As soon as they are saved to a variable, it's gone. Seems you can only do this with XS.
The reason for the readonly error https://github.com/ingydotnet/yaml-libyaml-pm/issues/25 is documented in perldoc perlguts under "AVs, HVs and undefined values"
